File: npm/typescript_oxlint/ts/rules/no_mixed_enums.ts
import { createNativeRule } from "./rule_creator";
import { isNumberLikeNode, isStringLikeNode } from "./type_utils";

export const noMixedEnumsRule = createNativeRule(
  "no-mixed-enums",
  {
    docs: {
      description: "Disallow mixing string and numeric enum members.",
    },
    messages: {
      mixed: "Mixing number and string enums can be confusing.",
    },
  },
  (context) => ({
    TSEnumDeclaration(node: any) {
      const members = enumMembersOf(node);
      if (members.length === 0) {
        return;
      }
      const desiredKind = enumMemberKind(context, members[0]);
      if (desiredKind === "unknown") {
        return;
      }
      for (const member of members) {
        const currentKind = enumMemberKind(context, member);
        if (currentKind === "unknown") {
          return;
        }
        if (currentKind !== desiredKind) {
          context.report({
            node: member.initializer ?? member,
            messageId: "mixed",
          });
          return;
        }
      }
    },
  }),
);

function enumMembersOf(node: any): readonly any[] {
  return node.body?.members ?? node.members ?? [];
}

function enumMemberKind(context: any, member: any): "number" | "string" | "unknown" {
  const initializer = member.initializer;
  if (!initializer) {
    return "number";
  }
  if (initializer.type === "Literal") {
    if (typeof initializer.value === "number") {
      return "number";
    }
    if (typeof initializer.value === "string") {
      return "string";
    }
  }
  if (isStringLikeNode(context, initializer)) {
    return "string";
  }
  if (isNumberLikeNode(context, initializer)) {
    return "number";
  }
  return "unknown";
}

File: src/core/corsa_core/src/lint/rules/no_mixed_enums.rs
use super::super::{LintNode, RuleContext, RuleMessage, RustLintRule};
use crate::lint::helpers::{EnumMemberKind, enum_member_kind, enum_members_of};

#[derive(Clone, Copy, Debug, Default)]
pub struct NoMixedEnumsRule;

const MESSAGES: &[RuleMessage] = &[RuleMessage {
    id: "mixed",
    description: "Mixing number and string enums can be confusing.",
}];
const LISTENERS: &[&str] = &["TSEnumDeclaration"];

impl RustLintRule for NoMixedEnumsRule {
    fn name(&self) -> &'static str {
        "no-mixed-enums"
    }

    fn docs_description(&self) -> &'static str {
        "Disallow mixing string and numeric enum members."
    }

    fn messages(&self) -> &'static [RuleMessage] {
        MESSAGES
    }

    fn listeners(&self) -> &'static [&'static str] {
        LISTENERS
    }

    fn check(&self, ctx: &mut RuleContext<'_>, node: &LintNode) {
        if node.kind != "TSEnumDeclaration" {
            return;
        }
        let members = enum_members_of(node);
        let Some(first_member) = members.first() else {
            return;
        };
        let desired_kind = enum_member_kind(first_member);
        if desired_kind == EnumMemberKind::Unknown {
            return;
        }
        for member in members {
            let current_kind = enum_member_kind(member);
            if current_kind == EnumMemberKind::Unknown {
                return;
            }
            if current_kind != desired_kind {
                ctx.report(
                    "mixed",
                    member
                        .child("initializer")
                        .map(|initializer| initializer.range)
                        .unwrap_or(member.range),
                );
                return;
            }
        }
    }
}
